### PR TITLE
fix nan handling in pow jvp (fixes #347)

### LINKED ISSUE #347: Issue taking gradients through np.where when one of branches is nan.
If you take gradients through np.where and one of the branches is nan, the gradient will be nan even if the branch is not taken. 

Here's a repro:

```
def test_fail(x):
  x = np.sqrt(np.sum(x ** 2, axis=1))
  ones = np.ones_like(x)
  x = np.where(x > 0.5, x, ones)
  return np.sum(x)

def test_succeed(x):
  x = np.sum(x ** 2, axis=1)
  ones = np.ones_like(x)
  x = np.sqrt(np.where(x > 0.5, x, ones))
  return np.sum(x)

x = np.array([[1, 2], [3, 4], [0, 0]], dtype=np.float64)

print('Test np.sqrt then np.where = {}'.format(test_fail(x)))
print('Gradient:')
print(grad(test_fail)(x))

print('Test np.where then np.sqrt = {}'.format(test_succeed(x)))
print('Gradient:')
print(grad(test_succeed)(x))
```

### PR DESCRIPTION
**Pull request by mattjj**
See #347 for example code.

Interestingly, Autograd still has this issue:

```python
import autograd.numpy as np
from autograd import grad

def f(x):
  x = np.sqrt(np.sum(x**2, axis=1))
  return np.sum(np.where(x > 0.5, x, np.ones_like(x)))

x = np.array([[1, 2], [3, 4], [0, 0]], dtype=np.float64)
print grad(f)(x)

# [[0.4472136  0.89442719]
#  [0.6        0.8       ]
#  [       nan        nan]]
```

It comes down to the formula for the jvp of `sqrt(x)` being problematic at `x=0` when the gradient has coefficients that are also zero, since `g * x ** -0.5` is `0 * inf` there. In this case, the zero coefficients in the gradient arise from the `lax.select`/`np.where` call.

I think a zero in the gradient should win out over an inf in the Jacobian, and this PR implements that policy. I had trouble implementing it on top of the standard `lax.mul` primitive because of linearity constraints in the JVP rule, so instead I added a new `lax._safe_mul` that implements the `0*inf = 0` policy in its translation rule.

### CODE REVIEW

**Comment by hawkinsp** [comment]
Travis is unhappy.

**Comment by mattjj** [comment]
Yeah sorry, this isn't ready for review yet!

File: jax/lax.py [hunk 1]
```diff
@@ -957,11 +960,15 @@ def _conj_transpose_rule(t, x, input_dtype):
...
 pow_p = standard_binop([_float | _complex, _float | _complex], 'pow')
-ad.defjvp(pow_p,
-          lambda g, x, y: mul(_brcast(g, y), mul(y, pow(x, select(
-              eq(y, _zeros(y)), _ones(y), sub(y, _ones(y)))))),
-          lambda g, x, y: mul(_brcast(g, x),
-                              mul(log(_replace_zero(x)), pow(x, y))))
+
+def pow_jvp_lhs(g, x, y):
+  jac = mul(y, pow(x, select(eq(y, _zeros(y)), _ones(y), sub(y, _ones(y)))))
+  return _safe_mul(_brcast(g, y), jac)
```
Comment (hawkinsp) [inline jax/lax.py]: Perhaps add a comment about the reason for the safe mul (or a link to the GH issue?) It's clearly subtle.

File: jax/lax.py [hunk 2]
```diff
@@ -992,6 +999,20 @@ def _add_transpose(t, x, y):
...
+  dtype = c.GetShape(x).numpy_dtype()
+  zero = c.Constant(onp.array(0, dtype=dtype))
+  out_shape = tuple(onp.maximum(c.GetShape(x).dimensions(),
+                                c.GetShape(y).dimensions()))
+  return c.Select(c.Or(c.Eq(x, zero), c.Eq(y, zero)),
+                  c.Broadcast(zero, out_shape),
+                  c.Mul(x, y))
+
+safe_mul_p = standard_binop([_num, _num], 'safe_mul',
+                            translation_rule=_safe_mul_translation_rule)
```
Comment (hawkinsp) [inline jax/lax.py]: Won't this need a JVP/transpose rule too?

**Review by hawkinsp — APPROVED** [review]

File: jax/lax.py (hunk 2 above)
Comment (mattjj) [reply jax/lax.py]: Actually the ad.defbilinear_broadcasting line after this one should handle that. But I’ll make sure it has the same rules set up as mul.

Comment (mattjj) [reply jax/lax.py]: I think by comparison to the definition of mul_p above we've got all the rules covered here. Thanks for checking!

File: jax/lax.py (hunk 1 above)
Comment (mattjj) [reply jax/lax.py]: Done!

**Pull request merged**